org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/PatternMatchingForInstanceofFixCore.java

```diff
@@ -142,55 +145,127 @@ public boolean visit(final InstanceofExpression visited) {
 
 				IfStatement ifStatement= (IfStatement) currentNode.getParent();
 
+				ResultCollector collector = new ResultCollector(visited);
 				if (isPositiveCaseToAnalyze) {
-					return maybeMatchPattern(visited, ifStatement.getThenStatement());
+					collector.collect(ifStatement.getThenStatement());
+				} else if (ifStatement.getElseStatement() != null) {
+					collector.collect(ifStatement.getElseStatement());
+				} else if (ASTNodes.fallsThrough(ifStatement.getThenStatement())) {
+					collector.collect(ASTNodes.getNextSibling(ifStatement));
 				}
 
-				if (ifStatement.getElseStatement() != null) {
-					return maybeMatchPattern(visited, ifStatement.getElseStatement());
+				if (collector.hasResult()) {
+					fResult.add(collector.build());
+					return false;
 				}
+				return true;
+			}
+
+			static class ResultCollector {
+				final InstanceofExpression visited;
+				SimpleName variableName;
+				final List<VariableDeclarationStatement> statementsToRemove = new ArrayList<>();
+				final List<VariableDeclarationStatement> statementsToConvert = new ArrayList<>();
 
-				if (ASTNodes.fallsThrough(ifStatement.getThenStatement())) {
-					return maybeMatchPattern(visited, ASTNodes.getNextSibling(ifStatement));
+				ResultCollector(InstanceofExpression visited) {
+					this.visited = visited;
 				}
 
-				return true;
-			}
+				public PatternMatchingForInstanceofFixOperation build() {
+					return new PatternMatchingForInstanceofFixOperation(visited, statementsToRemove, statementsToConvert, variableName);
+				}
 
-			private boolean maybeMatchPattern(final InstanceofExpression visited, final Statement conditionalStatements) {
-				List<Statement> statements= ASTNodes.asList(conditionalStatements);
+				private String getIdentifierName(Expression expression) {
+					if (expression instanceof SimpleName simpleName) {
+						return simpleName.getIdentifier();
+					}
+					return null;
+				}
 
-				if (!statements.isEmpty()) {
-					VariableDeclarationStatement variableDeclarationExpression= ASTNodes.as(statements.get(0), VariableDeclarationStatement.class);
-					VariableDeclarationFragment variableDeclarationFragment= ASTNodes.getUniqueFragment(variableDeclarationExpression);
+				boolean hasResult() {
+					return variableName != null && !statementsToRemove.isEmpty();
+				}
 
-					if (variableDeclarationFragment != null
-							&& Objects.equals(visited.getRightOperand().resolveBinding(), variableDeclarationExpression.getType().resolveBinding())) {
-						CastExpression castExpression= ASTNodes.as(variableDeclarationFragment.getInitializer(), CastExpression.class);
+				void addMatching(final VariableDeclarationStatement statementToRemove, final SimpleName expressionToMove, boolean toConvert) {
```

Comment:
Minor nit: rename statementToRemove to be "statement", rename expressionToMove to be "name".  Much simpler.